Improve threading and test suite

wzieba opened the pull request:
### Description

This PR aims to:
- add some improvements to threading by removing `runBlocking` call in `getAssignments` method
- improve the test suite (https://github.com/Automattic/Automattic-Tracks-Android/pull/239#discussion_r1732906646)
- fix a quick in public API design of requesting cache refresh when getting assignments

### To test
This PR introduces a robust test suite. Still, you can perform some smoke tests if you'd like following instructions from #241

wzieba committed:
Provide dispatchers and introduce `FileBasedCache#latest`

In order to remove `runBlocking` from `ExPlat`, we have to have a way to get value from cache in a way, that doesn't require sync file-operation. That's why I introduce `latest` field, which keeps the latest cache value in memory.

Providing dispatchers is considered a good practice and might come handy during tests

wzieba committed:
Update empty-cache test case to validate result, instead of mock invocation

wzieba committed:
Update stale-cache test case to validate result, instead of mock invocation

wzieba committed:
Update fresh-cache test case to validate result, instead of mock invocation

wzieba committed:
Refactor: extract similarities

Now, that we have few new/changed tests, I extract some similarities to make tests more readable

wzieba committed:
Update force refreshing on fresh cache test case to validate result, instead of mock invocation

wzieba committed:
This test is now covered by `force refreshing will fetch new assignments, even if cache is fresh`

wzieba committed:
Update clear-cache test to validate result instead of mock invocation

wzieba committed:
`ExPlat#getVariation` should not offer `shouldRefreshIfStale`

This is a confusing API, because, even if user requests a refresh, we don't return fresh data immediately, but only queue this to fetch. This is not something I'd expect as the SDK consumer.

Also, we already have specialized refresh methods. I don't think we need more.

I plan to remove `shouldRefreshIfStale` in another commit.

wzieba committed:
Update test case for multiple gets and cache refresh in between

I've added some comments explaining a story of this test, as I found it a little confusing when I started working on this.

wzieba committed:
These "if experiments empty" tests are not crucial business logic. I don't think their necessary (nothing bad will happen if the list will be empty, and the check itself is trivial), and they bring maintenance overhead at the same time.

wzieba committed:
Format lines

wzieba committed:
Cover the exception throwing in debug behavior for case of missing experiment

I'm not sure if it's good idea to test the debug code, but at the same time this test is tiny, so maybe let's keep it

wzieba committed:
Remove unused test methods and dependencies

wzieba committed:
Rename "dummy" to "test"

Personal preference, I'm not a fan of "dummy"

wzieba committed:
Move test objects to a single place for easier navigation

wzieba committed:
Use `TestScope` instead of a background scope

Fixing typo

wzieba committed:
Address ktlint code analysis

ParaskP7 commented on experimentation/build.gradle:
```diff
@@ -49,9 +49,6 @@ dependencies {
     testImplementation "junit:junit:$junitVersion"
     testImplementation "org.assertj:assertj-core:$assertjVersion"
-    testImplementation "org.mockito.kotlin:mockito-kotlin:$mockitoKotlinVersion"
-    testImplementation "org.mockito:mockito-inline:$mockitoInlineVersion"
```
Praise (❤️): Woohoo, gone, bye-bye, so-long, hasta-la-svista! 👋

ParaskP7 commented on experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt:
```diff
@@ -47,10 +48,7 @@ class ExPlat internal constructor(
      */
-    fun getVariation(
-        experiment: Experiment,
-        shouldRefreshIfStale: Boolean = false,
```
Think (🤔): Interestanto, I'll wait for how the APIs ends-up being in the end, also, we might want to update the KDoc for this function.

ParaskP7 replied:
UPDATE: I am seeing that the KDoc thingy is happening on https://github.com/Automattic/Automattic-Tracks-Android/pull/243/commits/990b3a1d78f2f138cafa05c839694c32695b639e , marking this as resolved.

ParaskP7 commented on experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt:
```diff
@@ -64,7 +62,7 @@ class ExPlat internal constructor(
         }
         return activeVariations.getOrPut(experimentIdentifier) {
-            getAssignments(if (shouldRefreshIfStale) IF_STALE else NEVER)
+            getAssignments(NEVER)
```
Minor (🔍): This can now become a one-liner.

ParaskP7 commented on experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt:
```diff
@@ -93,8 +91,7 @@ class ExPlat internal constructor(
 
     private fun getAssignments(refreshStrategy: RefreshStrategy): Assignments {
-        val cachedAssignments: Assignments =
-            runBlocking { repository.getCached() } ?: Assignments(emptyMap(), 0, 0)
```
Praise (❤️): Nicely done with the in-memory value, I like! 🥇

ParaskP7 commented on experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt:
```diff
@@ -93,8 +91,7 @@ class ExPlat internal constructor(
     private fun getAssignments(refreshStrategy: RefreshStrategy): Assignments {
-        val cachedAssignments: Assignments =
-            runBlocking { repository.getCached() } ?: Assignments(emptyMap(), 0, 0)
+        val cachedAssignments: Assignments = repository.getCached() ?: Assignments(emptyMap(), 0, -1)
```
Question (❓): What's with the `-1` change? 🤔

wzieba replied:
TLDR; I replaced this with `getAssignments` possible to return `null` now. I think it's much more readable now, and makes more sense: if there's no cache, it's better to return `null` than an `Assignment` with made up values. 98bba0a58977c64dac88429d5e1f07c3ad0d380a

Context here was that I wanted this Assignment to be `stale` when `Clock` is `0` (in tests, making things more readable). It was also ok for my thinking process to make `fetchedAt` value a minus value (that doesn't make sense) to signal that it's not a valid object. But as I moved to use `null`, this doesn't matter.

ParaskP7 replied:
Great, thanks for this change @wzieba , I agree, this seems much more robust to me! 💯

ParaskP7 commented on experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt:
```diff
@@ -124,11 +121,12 @@ class ExPlat internal constructor(
             experiments: Set<Experiment>,
             logger: ExperimentLogger,
             coroutineScope: CoroutineScope,
+            dispatcher: CoroutineDispatcher = Dispatchers.IO,
```
Suggestion (🔍): I suggest moving `dispatcher` and `coroutineScope` at the end of the function's signatures, them becoming the last parameters in it.

wzieba replied:
Added to: https://github.com/Automattic/Automattic-Tracks-Android/pull/243#discussion_r1738128236

ParaskP7 commented on experimentation/src/test/java/com/automattic/android/experimentation/local/FileBasedCacheTest.kt:
```diff
@@ -52,13 +54,13 @@ internal class FileBasedCacheTest {
-    private fun fileBasedCache() =
-        FileBasedCache(cacheDir = createTempDirectory().toFile())
+    private fun fileBasedCache(scope: TestScope) =
+        FileBasedCache(cacheDir = createTempDirectory().toFile(), dispatcher = StandardTestDispatcher(scope.testScheduler), scope = scope)
```
Minor (🔍): Consider making this a bit more readable:

```
    private fun fileBasedCache(scope: TestScope) = FileBasedCache(
        cacheDir = createTempDirectory().toFile(),
        dispatcher = StandardTestDispatcher(scope.testScheduler),
        scope = scope
    )
```

ParaskP7 commented on experimentation/src/test/java/com/automattic/android/experimentation/remote/ExperimentRestClientTest.kt:
```diff
@@ -13,24 +15,15 @@ import okhttp3.mockwebserver.RecordedRequest
 
 internal class ExperimentRestClientTest {
 
     private val server: MockWebServer = MockWebServer()
```
Praise (❤️): Love the new non-mocked tests! 🥇

ParaskP7 commented on experimentation/src/test/java/com/automattic/android/experimentation/ExPlatTest.kt:
```diff
@@ -2,267 +2,196 @@ package com.automattic.android.experimentation
+    fun `refreshing in case of fresh cache doesn't fetch new assignments`() = runTest {
+        enqueueSuccessfulNetworkResponse()
+        val exPlat = createExPlat(clock = { 3599 })
+        tempCache.saveAssignments(testAssignment.copy(timeToLive = 3600))
```
Praise (❤️): Nicely done!

ParaskP7 commented on experimentation/src/test/java/com/automattic/android/experimentation/ExPlatTest.kt:
```diff
@@ -2,267 +2,196 @@ package com.automattic.android.experimentation
+    as it could lead to unexpected behavior.
+     */
+    @Test
+    fun `getting variation for the second time returns the same value, even if cache was updated`() =
```
1. Praise (❤️): Thanks for this test, it helps further clarifying few API decisions.
2. Minor (🔍): Btw, you could move `runTest {` on the same line to the `fun` itself and make this test more readable, give it more horizontal space.

wzieba replied:
> Btw, you could move runTest { on the same line

I wish, but then auto-formatting moves it back. I didn't want to spend much time on fixing the rule, though

ParaskP7 replied:
Ah, I have a default line length set at 120 before hard wrapping, maybe that's why, yea, please ignore then... 🤷

ParaskP7 commented on experimentation/src/test/java/com/automattic/android/experimentation/ExPlatTest.kt:
```diff
@@ -2,267 +2,196 @@ package com.automattic.android.experimentation
+        val exPlat = createExPlat(experiments = emptySet())
 
-        verify(restClient, never()).fetchAssignments(eq(platform), any(), anyOrNull())
+        assertThatThrownBy {
```
Praise (❤️): TIL

ParaskP7 commented on experimentation/src/test/java/com/automattic/android/experimentation/ExPlatTest.kt:
```diff
@@ -2,267 +2,196 @@ package com.automattic.android.experimentation
-        setupAssignments(cachedAssignments = treatmentAssignments, fetchedAssignments = treatmentAssignments)
-
-        val secondVariation = exPlat.getVariation(dummyExperiment, shouldRefreshIfStale = false)
-        assertThat(secondVariation).isEqualTo(controlVariation)
```
Suggestion (💡): I usually split such code with an extra `/* HELPER */` grouping comment, which means, see helper functions below.

ParaskP7 reviewed (commented):
Same as with #239, #240 and #241 @wzieba , awesome work on the testing side of things, very helpful, and I'll continue reviewing the final PR, I am very close to the finish line, then we can start merging everything! 🤞

wzieba committed:
Merge branch 'experimentation_refactor' into improve_threading_and_test_suite

wzieba committed:
Make `getAssignments` private method return `null` if assignments are not found in cache

It's more reasonable to return `null` if there are no assignments, rather than a fake assignment with made up values.

wzieba committed:
Make `FileBasedCacheTest` more readable

wzieba commented:
@ParaskP7 I think this PR is ready to re-review. Re tests we discussed in https://github.com/Automattic/Automattic-Tracks-Android/pull/240#discussion_r1736152248 - I think I'll add it rather in #243 as there are some changes to tests there, so I don't want to work unnecessarily addressing conflicts

wzieba committed:
Address ktlint code analysis

ParaskP7 commented:
Reviewed and everything looks good to me, feel free to merge! ✅ 

PS: Just one note from me, maybe we should consider testing the publicly available `getVariation(...)` call, via the example app, as I was surprised that the change to making it nullable didn't have any effect their too... 🤔

wzieba commented:
Sorry, my bad. Actually `getVariation` is not changing signature - in case of empty assignment, we return `Control`

wzieba committed:
Fix `getVariation` signature

ParaskP7 commented:
> Sorry, my bad. Actually getVariation is not changing signature - in case of empty assignment, we return Control

Ah, right, true! 💯
